libkbfs/prefetcher.go

```diff
@@ -425,6 +456,69 @@ func (p *blockPrefetcher) handlePrefetch(pre *prefetch, isPrefetchNew,
 	return numBlocks, isTail, nil
 }
 
+func (p *blockPrefetcher) rescheduleTopBlock(
+	blockID kbfsblock.ID, pp *prefetch) {
+	if len(pp.parents) > 0 {
+		p.cancelPrefetch(blockID, pp)
+		return
+	}
+
+	delete(p.prefetches, blockID)
+	pp.Close()
+
+	// Only reschedule the top-most blocks, which has no parents.
+	rp, ok := p.rescheduled[blockID]
+	if !ok {
+		rp = &rescheduledPrefetch{
+			off: p.makeNewBackOff(),
+		}
+		p.rescheduled[blockID] = rp
+	}
+
+	if rp.timer != nil {
+		// Prefetch already scheduled.
+		return
+	}
+	// Copy the req but use a new empty block.
+	req := *pp.req
+	req.block = pp.req.block.NewEmpty()
```

Comment:
Actually here's another concern. If you're using a `NewEmpty` block, then when you go to reschedule this prefetch, then you'll actually have to _fetch_ it again (probably from the cache).

Right now the prefetcher assumes that if it encounters a `prefetchRequest` from `prefetchRequestCh`, then that means that the block is populated. So if you want to do `NewEmpty` here, then you'll need to fetch it.

Here with the rescheduling you're trying to simulate the scenario that a prefetch request didn't exist in the prefetcher before, but upon reschedule is treated as if it'd just gotten fetched. So there must be `Block` content.

I think the right way to do this is:

```
req := *pp.req
// At this point you are done with `pp`.
...
rp.timer = time.AfterFunc(d, func() {
	p.triggerPrefetch(&req)
})
```

Reply:
The way I think of it:
`prefetch` instances cached inside the prefetcher might not have a `*prefetchRequest` yet.
But a `prefetchRequest` should always have populated `Block` data.

Reply:
Sounds good thanks. Done in the latest commit.